aggregator/elem.go

```diff
@@ -0,0 +1,366 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3aggregator/aggregation"
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+var (
+	errCounterElemClosed = errors.New("counter element is closed")
+	errTimerElemClosed   = errors.New("timer element is closed")
+	errGaugeElemClosed   = errors.New("gauge element is closed")
+)
+
+type aggMetricFn func(
+	idPrefix []byte,
+	id metric.ID,
+	idSuffix []byte,
+	timestamp time.Time,
+	value float64,
+	policy policy.Policy,
+)
+
+type newMetricElemFn func() metricElem
+
+// metricElem is the common interface for metric elements
+type metricElem interface {
+	// ResetSetData resets the counter and sets data
+	ResetSetData(id metric.ID, policy policy.Policy)
+
+	// AddMetric adds a new metric value
+	// TODO(xichen): a value union would suffice here
+	AddMetric(timestamp time.Time, mu unaggregated.MetricUnion) error
+
+	// ReadAndDiscard processes values before a given time and discards
+	// them afterwards, returning whether the element can be collected
+	// after discarding the values
+	ReadAndDiscard(earlierThan time.Time, fn aggMetricFn) bool
+
+	// MarkAsTombstoned marks an element as tombstoned, which means this element
+	// will be deleted once its aggregated values have been flushed
+	MarkAsTombstoned()
+
+	// Close closes the element
+	Close()
+}
+
+type elemBase struct {
+	sync.Mutex
+
+	opts       Options
+	id         metric.ID
+	policy     policy.Policy
+	tombstoned bool
+	closed     bool
+}
+
+// CounterElem is the counter element
+type CounterElem struct {
+	elemBase
+
+	values map[time.Time]*aggregation.Counter
+}
+
+// TimerElem is the timer element
+type TimerElem struct {
+	elemBase
+
+	values map[time.Time]*aggregation.Timer
+}
+
+// GaugeElem is the gauge element
+type GaugeElem struct {
+	elemBase
+
+	values map[time.Time]*aggregation.Gauge
+}
+
+// ResetSetData resets the counter and sets data
+func (e *elemBase) ResetSetData(id metric.ID, policy policy.Policy) {
```

Comment:
We call `ResetSetData` as part of initializing the elem before putting it to use, so shouldn't need to lock